### PR TITLE
Data container improvements

### PR DESCRIPTION
**Pull request by philbucher**
This PR adds the tests for `DataContainer` as well as major improvements for the version with raw memory

### CODE REVIEW

File: co_sim_io/impl/data_container.hpp [hunk 1]
```diff
@@ -110,27 +114,30 @@ class DataContainerRawMemory : public DataContainer<TDataType>
...
-
-            *mppData = (TDataType *)malloc((NewSize)*sizeof(TDataType)); // TODO maybe use realloc? //also check if using "std::"
+        if (NewSize > mCapacity) { // only increase the capacity if too small => same behavior as std::vector
+            if (mCapacity == 0) {
+                *mppData = nullptr;
+            }
+            *mppData = (TDataType *)realloc(*mppData, (NewSize)*sizeof(TDataType));
             CO_SIM_IO_ERROR_IF_NOT(*mppData) << "Memory reallocation failed";
+
+            mCapacity = NewSize;
```
Comment (philbucher) [inline co_sim_io/impl/data_container.hpp:129]: @pooyan-dadvand now the behavior is exactly the same as for std::vector, no matter how it is resized

Comment (pooyan-dadvand) [reply co_sim_io/impl/data_container.hpp:129]: If I recall correctly, in many implementations of `std::vector`, if there is a need to increase the capacity, they increase it to the twice of current size for efficiency reasons.

Comment (pooyan-dadvand) [reply co_sim_io/impl/data_container.hpp:129]: To be more precise: it will resize to the `max(mCapacity*2, NewSize)`

Comment (philbucher) [reply co_sim_io/impl/data_container.hpp:129]: good point!
I will check if it is possible also for us and implement it :)

Comment (philbucher) [reply co_sim_io/impl/data_container.hpp]: I did it a bit simpler, but can easily be changed in the future

Comment (philbucher) [reply co_sim_io/impl/data_container.hpp]: edit I now do what you suggested, seems to work better

**Comment by philbucher** [comment]
another thing finished before the vacation :D

**Pull request merged**